# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)


children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Для этого найдем отношение фактов задолжности по кредиту к заёмщикам, имеющим разное количество детей.

In [22]:
#Создадим новую таблицу, для сохранения результатов расчета

children_relation = pd.DataFrame({'children':[0, 1, 2, 3, 4, 5]})

Напишем цикл, который будет принимать уникальные данные по кол-ву детей из исходной таблицы
и расчитывать отношение кол-ва задолжностей к заёмщикам с разным кол-вом детей.
Результат сохраним в столбце relation, новой таблицы.

In [23]:
for ch in data['children'].sort_values().unique():
    children_relation.loc[children_relation['children']==ch,'relation'] = \
    data.loc[data['children']==ch]['debt'].sum() / data.loc[data['children']==ch]['children'].count()

Для удобства, переведем значения стобца relation в проценты

In [24]:
children_relation['relation'] = children_relation['relation'].apply(lambda x: format(x, '.2%'))

In [25]:
#Выдедем таблицу с результатом расчетов

display(children_relation.sort_values(by='relation'))

children relation
5         5    0.00%
0         0    7.54%
3         3    8.18%
1         1    9.23%
2         2    9.45%
4         4    9.76%

Из результатов видно: количество заемщиков с задолжнотью, имеющих одного, **двух или четырех детей** похоже (9.23%, 9.45%, 9.76%). При этом разница между заёмщиками, **не имеющими детей** и заёмщиками **с одним, двумя, четыремя детьми** более 1%. 

**Вывод:** Количество детей напрямую не влияет на вылату заёмщиком кредита в срок. Однако заёмщики у которых **нет детей**, чаще не имеют задолжность по кредиту.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Для этого определим отношение заёмщиков с задолжностью к общему количесту заёмщиков, с разным семейным положением

In [26]:
#Создадим новую таблицу для визуализации расчётов 
family_relation = pd.DataFrame({'family_status':['Не женат / не замужем','в разводе','вдовец / вдова','гражданский брак',
 'женат / замужем']})

Напишем цикл, принимающий количество заёмщиков по их семейному положению.
И найдем отношение заёмщиков с задолжностью к общему кол-ву заёмщиков по семейному положению.
Добавим результат в столбец relation.

In [27]:
for fam in data['family_status'].sort_values().unique():
    family_relation.loc[family_relation['family_status']==fam,'relation'] = \
    data.loc[data['family_status']==fam]['debt'].sum() / data.loc[data['family_status']==fam]['family_status'].count()

Для удобства переведём результат в проценты

In [28]:
family_relation['relation'] = family_relation['relation'].apply(lambda x: format(x, '.2%'))

In [29]:
#Выведем таблицу с результатом расчётов 
display(family_relation.sort_values(by='relation'))

family_status relation
2         вдовец / вдова    6.62%
1              в разводе    7.06%
4        женат / замужем    7.56%
3       гражданский брак    9.31%
0  Не женат / не замужем    9.76%

По результатам видно: заёмщики со статусами **вдовец / вдова, в разводе, женат / замужем** реже имеют задолжности по кредиту, в отличие от заёмщиков с статусом **Не женат / не замужем**.

**Вывод:** Семейное положение влияет на возврат кредита в срок. Среди заёмщиков со стасум **Не женат / не замужем** значительно чаще встречаются задолжности по кредиту, чем у заёмщиков со статусом **женат / замужем** (разница 2.2%).

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Для этого необходимо узнать отношение заёмщиков с задолжностью по кредиту к общему количеству заёмщиков, с разным доходом.

In [30]:
# Создадим новую таблицу для визуализации результатов

income_relation = pd.DataFrame({'total_income':['более 1000001', 'от 200001 до 1000000', 'от 50001 до 200000', \
                                                'от 30001 до 50000','менее 30000'],
                               'category':['A', 'B', 'C', 'D', 'E']})

Напишем цикл, который будет принимать уникальное кол-во заёмщиков, по категориям доходов
и находить отношение заёмщиков с задолжностью, к общему кол-ву заёмщиков.

In [31]:
for inc in data['total_income_category'].sort_values().unique():
    income_relation.loc[income_relation['category']==inc,'relation'] = \
    data.loc[data['total_income_category']==inc]['debt'].sum() \
    / data.loc[data['total_income_category']==inc]['total_income_category'].count()

Для удобства переведём резутат в проценты

In [32]:
income_relation['relation'] = income_relation['relation'].apply(lambda x: format(x, '.2%'))

In [33]:
#Выведем таблицу с результатми
display(income_relation.sort_values(by='relation'))

total_income category relation
3     от 30001 до 50000        D    6.02%
1  от 200001 до 1000000        B    7.06%
0         более 1000001        A    8.00%
2    от 50001 до 200000        C    8.50%
4           менее 30000        E    9.09%

По таблице видно: среди заёмщиков, с зарплатой **менее 30 000** больше всего просрочек по кредиту. Заёмщики же с зарплатой **от 30 000 до 50 000** реже имеют задолжности по кредиту.

**Вывод:** Уровень дохода влияет на факт возврата кредита в срок. Среди заёмщиков с уровнем дохода **менее 30 000** больший процент задолжностей по кредиту. В отличие от заёмщиков с доходом **от 30 000 до 50 000**, среди них процент задолжностей на 3% меньше. 

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Для этого получим отношение заёмщиков по категориям причин получерия кредита к общему количеству заемщиков по этим категориям.

In [34]:
# Создадим новую таблицу для вычислений
purpose_relation = pd.DataFrame({'category':['операции с автомобилем', 'операции с недвижимостью', \
                                             'получение образования', 'проведение свадьбы']})

Необходимо создать цикл, который будет принимать на вход количество заёмщиков с задолжностью
по уникальным категориям причин получения кредитаи найти их отношщение к общему кол-ву заёмщиков по этим категориям.

In [35]:
for pur in data['purpose_category'].sort_values().unique():
    purpose_relation.loc[purpose_relation['category']==pur,'relation'] = \
    data.loc[data['purpose_category']==pur]['debt'].sum() \
    / data.loc[data['purpose_category']==pur]['purpose_category'].count()

Для удобства переведём результат в проценты

In [36]:
purpose_relation['relation'] = purpose_relation['relation'].apply(lambda x: format(x, '.2%'))

In [37]:
#Выведем таблицу с вычислениями
display(purpose_relation.sort_values(by='relation'))

category relation
1  операции с недвижимостью    7.26%
3        проведение свадьбы    7.91%
2     получение образования    9.25%
0    операции с автомобилем    9.35%

Из таблицы с вычислениями видно: заёмщики, которые берут кредит **операции с недвижимостью** и **проведение свадьбы** имеют меньший процент задолжностей, в отличие от заёмщиков с причиной получения кредита **операции с автомобилем** и **получение образования**(разница между ними в среднем ~ 1%)

**Вывод:** Цели кредита по-разному влияют на факт выплаты кредита в срок. Среди заёмщиков, берущих кредит на **получение образования** и **операции с автомобилем** реже встречаются факты просрочки погашения кредита.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски в исходных данных были в стобцах **days_employed — общий трудовой стаж в днях** и **total_income — ежемесячный доход**, они могут быть связаны с отсутвием информации от самих заёмщмков или ошибкой сотрудников, заполняющих анкеты для заёмщмков. Так же данные могли быть утеряны, при переносе их в другое хранилище или при их первичной обратотке.

В исходной таблице пропущены значения в стобцах **days_employed — общий трудовой стаж в днях** и **total_income — ежемесячный доход** в одинаковых строчках и одикакое количество (2174). Вероятнее всего это означает безработных заёмщмиков. У них соответственно нет трудового стажа и дохода. 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Количественные переменные в данных могут имень большую разницу между значениями, поэтому заполнять пропуски средним значением в таких данных некорректно. Среднее значение будет сильно зависеть от максимального и минимального значения. 

### Шаг 4: общий вывод.

#### Мы рассмотрели данные и смогли ответить на вопросы о зависимости выплаты кредита в срок у разных заёмщиков.

**Количество детей:**


Заёмщики, **не имеющие детей**, значительно реже имеют задолжности по выплате кредита, в отличие от заёмщиков, **имеющих детей**(кол-во детей незначительно влияет на факт выплаты кредита в срок).

**Семейное положение:**

Среди заёмщиков с семейным статусом **женат / замужем** меньшее количество задолжностей по кредиту, чем у заёмщиков с статусом **Не женат / не замужем** наибольший процент просрочек по кредиту. Так же наименьшее кол-во просрочек, среди заёмщиков с статусом **вдовец / вдова**, **в разводе**.

**Уровень дохода:**

Наиболее активно возвращают кредит заёмщики с доходом **от 30 000 до 50 000**, в отличие от заёмщиков с доходом **менее 30 000**(среди них наибольший процент задолжности по кредиту)

**Цели кредита:**

Заёмщики с целями **операции с недвижимостью** и **проведение свадьбы** имеют меньший процент задолжностей по кредиту, а вот заёмщики с целями **получение образования** и **операции с автомобилем** имеют наибольший процент фактов задолжноти по кредиту.	

**Результаты исследования можно использовать для построения модели *кредитного скоринга* и уточнения критериев способности потенциального заёмщика вернуть дегьги банку.**